{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MyVariant.info and MyGene.info Use Case"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following R script demonstrates the utility of MyVariant.info and MyGene.info R clients to annotate variants and prioritize candidate genes in patients with rare Mendelian diseases. This specific study uses data obtained from the database of phenotype and genotype (dbGaP) study...FASTQ files generated by Ng et al for the Miller syndrome study. FASTQ files were processed according to the Broad Institute’s best practices. Individual samples were aligned to the hg19 reference genome using BWA-MEM 0.7.10. Variants were called using GATK 3.3-0 HaplotypeCaller and quality scores were recalibrated using GATK VariantRecalibrator."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: GenomicFeatures\n",
      "Loading required package: AnnotationDbi\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘AnnotationDbi’\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "Attaching package: ‘mygene’\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    query\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(myvariant)\n",
    "library(mygene)\n",
    "library(VariantAnnotation)\n",
    "source(\"https://raw.githubusercontent.com/Network-of-BioThings/myvariant.info/master/docs/ipynb/mendelian.R\")\n",
    "setwd(\"~/sulab/myvariant/vcf/\")\n",
    "vcf.files <- paste(getwd(), list.files(getwd()), sep=\"/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "mendelian.R defines some helper functions that are used in the analysis downstream of retrieving annotations:\n",
    "\n",
    "`replaceWith0` - replaces all NAs with 0.\n",
    "\n",
    "`rankByCaddScore` - for prioritizing genes by deleteriousness (scaled CADD score)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Annotating variants with MyVariant.info\n",
    "The following function reads in each output VCF file using the VariantAnnotation package available from Bioconductor. Install with `biocLite(\"VariantAnnotation\")`. `formatHgvs` (from the ***myvariant*** Bioconductor package) is a function that reads the genomic location and variant information from the VCF to create HGVS IDs which serve as a primary key for each variant. The function `getVariants` makes the queries to MyVariant.info to retrieve annotations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing  /Users/Adam/sulab/myvariant/vcf/subject01_recalibrate_SNP.vcf ...\n",
      "found header lines for 3 ‘fixed’ fields: ALT, QUAL, FILTER \n",
      "found header lines for 24 ‘info’ fields: AC, AF, AN, BaseQRankSum, ClippingRankSum, DB, DP, DS, END, FS, HaplotypeScore, InbreedingCoeff, MLEAC, MLEAF, MQ, MQ0, MQRankSum, NEGATIVE_TRAIN_SITE, POSITIVE_TRAIN_SITE, QD, ReadPosRankSum, SOR, VQSLOD, culprit \n",
      "found header lines for 5 ‘geno’ fields: GT, AD, DP, GQ, PL \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Querying chunk 1 of 40\n",
      "Querying chunk 2 of 40\n",
      "Querying chunk 3 of 40\n",
      "Querying chunk 4 of 40\n",
      "Querying chunk 5 of 40\n",
      "Querying chunk 6 of 40\n",
      "Querying chunk 7 of 40\n",
      "Querying chunk 8 of 40\n",
      "Querying chunk 9 of 40\n",
      "Querying chunk 10 of 40\n",
      "Querying chunk 11 of 40\n",
      "Querying chunk 12 of 40\n",
      "Querying chunk 13 of 40\n",
      "Querying chunk 14 of 40\n",
      "Querying chunk 15 of 40\n",
      "Querying chunk 16 of 40\n",
      "Querying chunk 17 of 40\n",
      "Querying chunk 18 of 40\n",
      "Querying chunk 19 of 40\n",
      "Querying chunk 20 of 40\n",
      "Querying chunk 21 of 40\n",
      "Querying chunk 22 of 40\n",
      "Querying chunk 23 of 40\n",
      "Querying chunk 24 of 40\n",
      "Querying chunk 25 of 40\n",
      "Querying chunk 26 of 40\n",
      "Querying chunk 27 of 40\n",
      "Querying chunk 28 of 40\n",
      "Querying chunk 29 of 40\n",
      "Querying chunk 30 of 40\n",
      "Querying chunk 31 of 40\n",
      "Querying chunk 32 of 40\n",
      "Querying chunk 33 of 40\n",
      "Querying chunk 34 of 40\n",
      "Querying chunk 35 of 40\n",
      "Querying chunk 36 of 40\n",
      "Querying chunk 37 of 40\n",
      "Querying chunk 38 of 40\n",
      "Querying chunk 39 of 40\n",
      "Querying chunk 40 of 40\n",
      "Concatenating data, please be patient.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing  /Users/Adam/sulab/myvariant/vcf/subject02_recalibrate_SNP.vcf ...\n",
      "found header lines for 3 ‘fixed’ fields: ALT, QUAL, FILTER \n",
      "found header lines for 24 ‘info’ fields: AC, AF, AN, BaseQRankSum, ClippingRankSum, DB, DP, DS, END, FS, HaplotypeScore, InbreedingCoeff, MLEAC, MLEAF, MQ, MQ0, MQRankSum, NEGATIVE_TRAIN_SITE, POSITIVE_TRAIN_SITE, QD, ReadPosRankSum, SOR, VQSLOD, culprit \n",
      "found header lines for 5 ‘geno’ fields: GT, AD, DP, GQ, PL \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Querying chunk 1 of 39\n",
      "Querying chunk 2 of 39\n",
      "Querying chunk 3 of 39\n",
      "Querying chunk 4 of 39\n",
      "Querying chunk 5 of 39\n",
      "Querying chunk 6 of 39\n",
      "Querying chunk 7 of 39\n",
      "Querying chunk 8 of 39\n",
      "Querying chunk 9 of 39\n",
      "Querying chunk 10 of 39\n",
      "Querying chunk 11 of 39\n",
      "Querying chunk 12 of 39\n",
      "Querying chunk 13 of 39\n",
      "Querying chunk 14 of 39\n",
      "Querying chunk 15 of 39\n",
      "Querying chunk 16 of 39\n",
      "Querying chunk 17 of 39\n",
      "Querying chunk 18 of 39\n",
      "Querying chunk 19 of 39\n",
      "Querying chunk 20 of 39\n",
      "Querying chunk 21 of 39\n",
      "Querying chunk 22 of 39\n",
      "Querying chunk 23 of 39\n",
      "Querying chunk 24 of 39\n",
      "Querying chunk 25 of 39\n",
      "Querying chunk 26 of 39\n",
      "Querying chunk 27 of 39\n",
      "Querying chunk 28 of 39\n",
      "Querying chunk 29 of 39\n",
      "Querying chunk 30 of 39\n",
      "Querying chunk 31 of 39\n",
      "Querying chunk 32 of 39\n",
      "Querying chunk 33 of 39\n",
      "Querying chunk 34 of 39\n",
      "Querying chunk 35 of 39\n",
      "Querying chunk 36 of 39\n",
      "Querying chunk 37 of 39\n",
      "Querying chunk 38 of 39\n",
      "Querying chunk 39 of 39\n",
      "Concatenating data, please be patient.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing  /Users/Adam/sulab/myvariant/vcf/subject03_recalibrate_SNP.vcf ...\n",
      "found header lines for 3 ‘fixed’ fields: ALT, QUAL, FILTER \n",
      "found header lines for 24 ‘info’ fields: AC, AF, AN, BaseQRankSum, ClippingRankSum, DB, DP, DS, END, FS, HaplotypeScore, InbreedingCoeff, MLEAC, MLEAF, MQ, MQ0, MQRankSum, NEGATIVE_TRAIN_SITE, POSITIVE_TRAIN_SITE, QD, ReadPosRankSum, SOR, VQSLOD, culprit \n",
      "found header lines for 5 ‘geno’ fields: GT, AD, DP, GQ, PL \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Querying chunk 1 of 31\n",
      "Querying chunk 2 of 31\n",
      "Querying chunk 3 of 31\n",
      "Querying chunk 4 of 31\n",
      "Querying chunk 5 of 31\n",
      "Querying chunk 6 of 31\n",
      "Querying chunk 7 of 31\n",
      "Querying chunk 8 of 31\n",
      "Querying chunk 9 of 31\n",
      "Querying chunk 10 of 31\n",
      "Querying chunk 11 of 31\n",
      "Querying chunk 12 of 31\n",
      "Querying chunk 13 of 31\n",
      "Querying chunk 14 of 31\n",
      "Querying chunk 15 of 31\n",
      "Querying chunk 16 of 31\n",
      "Querying chunk 17 of 31\n",
      "Querying chunk 18 of 31\n",
      "Querying chunk 19 of 31\n",
      "Querying chunk 20 of 31\n",
      "Querying chunk 21 of 31\n",
      "Querying chunk 22 of 31\n",
      "Querying chunk 23 of 31\n",
      "Querying chunk 24 of 31\n",
      "Querying chunk 25 of 31\n",
      "Querying chunk 26 of 31\n",
      "Querying chunk 27 of 31\n",
      "Querying chunk 28 of 31\n",
      "Querying chunk 29 of 31\n",
      "Querying chunk 30 of 31\n",
      "Querying chunk 31 of 31\n",
      "Concatenating data, please be patient.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing  /Users/Adam/sulab/myvariant/vcf/subject04_recalibrate_SNP.vcf ...\n",
      "found header lines for 3 ‘fixed’ fields: ALT, QUAL, FILTER \n",
      "found header lines for 24 ‘info’ fields: AC, AF, AN, BaseQRankSum, ClippingRankSum, DB, DP, DS, END, FS, HaplotypeScore, InbreedingCoeff, MLEAC, MLEAF, MQ, MQ0, MQRankSum, NEGATIVE_TRAIN_SITE, POSITIVE_TRAIN_SITE, QD, ReadPosRankSum, SOR, VQSLOD, culprit \n",
      "found header lines for 5 ‘geno’ fields: GT, AD, DP, GQ, PL \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Querying chunk 1 of 38\n",
      "Querying chunk 2 of 38\n",
      "Querying chunk 3 of 38\n",
      "Querying chunk 4 of 38\n",
      "Querying chunk 5 of 38\n",
      "Querying chunk 6 of 38\n",
      "Querying chunk 7 of 38\n",
      "Querying chunk 8 of 38\n",
      "Querying chunk 9 of 38\n",
      "Querying chunk 10 of 38\n",
      "Querying chunk 11 of 38\n",
      "Querying chunk 12 of 38\n",
      "Querying chunk 13 of 38\n",
      "Querying chunk 14 of 38\n",
      "Querying chunk 15 of 38\n",
      "Querying chunk 16 of 38\n",
      "Querying chunk 17 of 38\n",
      "Querying chunk 18 of 38\n",
      "Querying chunk 19 of 38\n",
      "Querying chunk 20 of 38\n",
      "Querying chunk 21 of 38\n",
      "Querying chunk 22 of 38\n",
      "Querying chunk 23 of 38\n",
      "Querying chunk 24 of 38\n",
      "Querying chunk 25 of 38\n",
      "Querying chunk 26 of 38\n",
      "Querying chunk 27 of 38\n",
      "Querying chunk 28 of 38\n",
      "Querying chunk 29 of 38\n",
      "Querying chunk 30 of 38\n",
      "Querying chunk 31 of 38\n",
      "Querying chunk 32 of 38\n",
      "Querying chunk 33 of 38\n",
      "Querying chunk 34 of 38\n",
      "Querying chunk 35 of 38\n",
      "Querying chunk 36 of 38\n",
      "Querying chunk 37 of 38\n",
      "Querying chunk 38 of 38\n",
      "Concatenating data, please be patient.\n"
     ]
    }
   ],
   "source": [
    "getVars <- function(vcf.file){\n",
    "  cat(paste(\"Processing \", vcf.file, \"...\\n\", sep=\" \"))\n",
    "  vcf <- readVcf(vcf.file, genome=\"hg19\")\n",
    "  vcf <- vcf[isSNV(vcf)]\n",
    "  vars <- rowRanges(vcf)\n",
    "  vars <- as(vars, \"DataFrame\")\n",
    "  vars$query <- formatHgvs(vcf, \"snp\")\n",
    "  annotations <- getVariants(vars$query)\n",
    "  annotations[c('DP', 'FS', 'QD')] <- info(vcf)[c('DP', 'FS', 'QD')]\n",
    "  annotations <- replaceWith0(annotations)\n",
    "  annotations\n",
    "}\n",
    "\n",
    "vars <- lapply(vcf.files, getVars)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We then filter each of the VCF files based on coverage and annotations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2589 common genes are mutated amongst patients"
     ]
    }
   ],
   "source": [
    "cat(paste(nrow(subset(data.frame(table(unlist(lapply(vars, function(i) unique(i$dbnsfp.genename))))), \n",
    "    Freq == 4)), \"common genes are mutated amongst patients\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Filtering Steps\n",
    "1) The first filter utilizes annotations output by HaplotypeCaller from GATK."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2420 genes remain after filtering for coverage"
     ]
    }
   ],
   "source": [
    "filter1 <- lapply(vars, function(i) subset(i, DP > 8 & FS < 30 & QD > 2))\n",
    "    \n",
    "cat(paste(nrow(subset(data.frame(table(unlist(lapply(filter1, function(i) unique(i$dbnsfp.genename))))), \n",
    "    Freq == 4)), \"genes remain after filtering for coverage and strand bias\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2) The next filter keeps nonsynoymous and splite site variants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2003 genes remain after filtering for nonsynonymous and splice site variants"
     ]
    }
   ],
   "source": [
    "filter2 <- lapply(filter1, function(i) subset(i, cadd.consequence %in% c(\"NON_SYNONYMOUS\", \"STOP_GAINED\", \"STOP_LOST\", \n",
    "                                           \"CANONICAL_SPLICE\", \"SPLICE_SITE\")))\n",
    "    \n",
    "cat(paste(nrow(subset(data.frame(table(unlist(lapply(filter2, function(i) unique(i$dbnsfp.genename))))), \n",
    "    Freq == 4)), \"genes remain after filtering for nonsynonymous and splice site variants\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3) The third filter keeps rare variants according to the ExAC data set with allele frequency < 0.01."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28 genes remain after filtering for allele frequency < 0.01 in the ExAC dataset"
     ]
    }
   ],
   "source": [
    "filter3 <- lapply(filter2, function(i) subset(i, exac.af < 0.01))\n",
    "    \n",
    "cat(paste(nrow(subset(data.frame(table(unlist(lapply(filter3, function(i) unique(i$dbnsfp.genename))))), \n",
    "    Freq == 4)), \"genes remain after filtering for allele frequency < 0.01 in the ExAC dataset\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4) The last filter keeps rare variants according to the 1000 Genomes Project with allele frequency < 0.01."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14 genes remain after filtering for allele frequency < 0.01 in 1000 Genomes Project"
     ]
    }
   ],
   "source": [
    "filter4 <- lapply(filter3, function(i) subset(i, sapply(dbnsfp.1000gp1.af, function(j) j < 0.01 )))\n",
    "    \n",
    "top.genes <- subset(data.frame(table(unlist(lapply(filter4, function(i) unique(i$dbnsfp.genename))))), Freq == 4)\n",
    "top.genes <- subset(top.genes, !(Var1 %in% c(\"NULL\", 0)))\n",
    "cat(paste(nrow(top.genes), \"genes remain after filtering for allele frequency < 0.01 in 1000 Genomes Project\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>CDC27</li>\n",
       "\t<li>CDON</li>\n",
       "\t<li>CTBP2</li>\n",
       "\t<li>CTDSP2</li>\n",
       "\t<li>DHODH</li>\n",
       "\t<li>FAM104B</li>\n",
       "\t<li>FRG2C</li>\n",
       "\t<li>GXYLT1</li>\n",
       "\t<li>HYDIN</li>\n",
       "\t<li>KIR3DL3</li>\n",
       "\t<li>MTCH2</li>\n",
       "\t<li>POTED</li>\n",
       "\t<li>SETD8</li>\n",
       "\t<li>XKR3</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item CDC27\n",
       "\\item CDON\n",
       "\\item CTBP2\n",
       "\\item CTDSP2\n",
       "\\item DHODH\n",
       "\\item FAM104B\n",
       "\\item FRG2C\n",
       "\\item GXYLT1\n",
       "\\item HYDIN\n",
       "\\item KIR3DL3\n",
       "\\item MTCH2\n",
       "\\item POTED\n",
       "\\item SETD8\n",
       "\\item XKR3\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. CDC27\n",
       "2. CDON\n",
       "3. CTBP2\n",
       "4. CTDSP2\n",
       "5. DHODH\n",
       "6. FAM104B\n",
       "7. FRG2C\n",
       "8. GXYLT1\n",
       "9. HYDIN\n",
       "10. KIR3DL3\n",
       "11. MTCH2\n",
       "12. POTED\n",
       "13. SETD8\n",
       "14. XKR3\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] CDC27   CDON    CTBP2   CTDSP2  DHODH   FAM104B FRG2C   GXYLT1  HYDIN  \n",
       "[10] KIR3DL3 MTCH2   POTED   SETD8   XKR3   \n",
       "2141 Levels: 0 AAED1 ABCA12 ABCA2 ABCA6 ABCB10 ABCB4 ABCB6 ABCB7 ... ZZEF1"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top.genes$Var1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Prioritizing genes\n",
    "We can prioritize the genes that contain variants in each of the patients according to CADD (deleteriousness) score. `rankByCaddScore` extracts the average CADD scores of the variants in each gene per patient and ranks in descending order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>gene</th><th scope=col>cadd.phred</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>DHODH</td><td>26.81</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>GXYLT1</td><td>26.1</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>CTBP2</td><td>26</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>MTCH2</td><td>24.93333</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>SETD8</td><td>23.8</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>CTDSP2</td><td>22.58571</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>CDC27</td><td>20.465</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>HYDIN</td><td>18.91</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>XKR3</td><td>15.72</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>CDON</td><td>10.02</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>POTED</td><td>8.947</td></tr>\n",
       "\t<tr><th scope=row>12</th><td>FAM104B</td><td>7.2055</td></tr>\n",
       "\t<tr><th scope=row>13</th><td>FRG2C</td><td>1.883</td></tr>\n",
       "\t<tr><th scope=row>14</th><td>KIR3DL3</td><td>1.3765</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ll}\n",
       "  & gene & cadd.phred\\\\\n",
       "\\hline\n",
       "\t1 & DHODH & 26.81\\\\\n",
       "\t2 & GXYLT1 & 26.1\\\\\n",
       "\t3 & CTBP2 & 26\\\\\n",
       "\t4 & MTCH2 & 24.93333\\\\\n",
       "\t5 & SETD8 & 23.8\\\\\n",
       "\t6 & CTDSP2 & 22.58571\\\\\n",
       "\t7 & CDC27 & 20.465\\\\\n",
       "\t8 & HYDIN & 18.91\\\\\n",
       "\t9 & XKR3 & 15.72\\\\\n",
       "\t10 & CDON & 10.02\\\\\n",
       "\t11 & POTED & 8.947\\\\\n",
       "\t12 & FAM104B & 7.2055\\\\\n",
       "\t13 & FRG2C & 1.883\\\\\n",
       "\t14 & KIR3DL3 & 1.3765\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "      gene cadd.phred\n",
       "1    DHODH   26.81000\n",
       "2   GXYLT1   26.10000\n",
       "3    CTBP2   26.00000\n",
       "4    MTCH2   24.93333\n",
       "5    SETD8   23.80000\n",
       "6   CTDSP2   22.58571\n",
       "7    CDC27   20.46500\n",
       "8    HYDIN   18.91000\n",
       "9     XKR3   15.72000\n",
       "10    CDON   10.02000\n",
       "11   POTED    8.94700\n",
       "12 FAM104B    7.20550\n",
       "13   FRG2C    1.88300\n",
       "14 KIR3DL3    1.37650"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ranked <- rankByCaddScore(top.genes$Var1, filter4)\n",
    "ranked"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "##Annotating candidate genes with MyGene.info\n",
    "Now we can annotate the genes using the `queryMany` function from ***mygene*** which can be install from Bioconductor (`biocLite(\"mygene\")`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "DataFrame with 14 rows and 7 columns\n",
       "            MIM\n",
       "    <character>\n",
       "1        126064\n",
       "2        613321\n",
       "3        602619\n",
       "4        613221\n",
       "5        607240\n",
       "...         ...\n",
       "10       608707\n",
       "11       607549\n",
       "12           NA\n",
       "13           NA\n",
       "14       610095\n",
       "                                                                                 name\n",
       "                                                                          <character>\n",
       "1                                              dihydroorotate dehydrogenase (quinone)\n",
       "2                                                      glucoside xylosyltransferase 1\n",
       "3                                                        C-terminal binding protein 2\n",
       "4                                                             mitochondrial carrier 2\n",
       "5                                  SET domain containing (lysine methyltransferase) 8\n",
       "...                                                                               ...\n",
       "10                                       cell adhesion associated, oncogene regulated\n",
       "11                                               POTE ankyrin domain family, member D\n",
       "12                                      family with sequence similarity 104, member B\n",
       "13                                                FSHD region gene 2 family, member C\n",
       "14  killer cell immunoglobulin-like receptor, three domains, long cytoplasmic tail, 3\n",
       "    pathway.kegg uniprot.Swiss-Prot uniprot.TrEMBL       query         _id\n",
       "          <List>        <character>         <List> <character> <character>\n",
       "1       ########             Q02127       ########       DHODH        1723\n",
       "2       ########             Q4G148       ########      GXYLT1      283464\n",
       "3       ########             P56545       ########       CTBP2        1488\n",
       "4       ########             Q9Y6C9       ########       MTCH2       23788\n",
       "5       ########             Q9NQR1       ########       SETD8      387893\n",
       "...          ...                ...            ...         ...         ...\n",
       "10      ########             Q4KMG0       ########        CDON       50937\n",
       "11      ########             Q86YR6       ########       POTED      317754\n",
       "12      ########             Q5XKR9       ########     FAM104B       90736\n",
       "13      ########             A6NGY1       ########       FRG2C   100288801\n",
       "14      ########                 NA       ########     KIR3DL3      115653"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "queryMany(ranked$gene, scopes=\"symbol\", species=\"human\", fields=c(\"name\", \"pathway.kegg\", \"MIM\", \"uniprot\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Conclusion\n",
    "Ng et al concluded DHODH is responsible for Miller Syndrome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.2.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
