{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script demonstrates the utility of MyVariant.info and MyGene.info to annotate variants and prioritize candidate genes in patients with rare Mendelian diseases. This specific study uses data obtained from the database of phenotype and genotype (dbGaP) study...FASTQ files generated by Ng et al for the Miller syndrome study. FASTQ files were processed according to the Broad Institute’s best practices. Individual samples were aligned to the hg19 reference genome using BWA-MEM 0.7.10. Variants were called using GATK 3.3-0 HaplotypeCaller and quality scores were recalibrated using GATK VariantRecalibrator."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "library(myvariant)\n",
    "library(mygene)\n",
    "library(VariantAnnotation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "setwd(\"~/sulab/myvariant/vcf/\")\n",
    "vcf.files <- paste(getwd(), list.files(getwd()), sep=\"/\")\n",
    "source(\"https://raw.githubusercontent.com/Network-of-BioThings/myvariant.info/master/docs/ipynb/mendelian.R\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "mendelian.R defines some helper functions that are used in the analysis downstream of retrieving annotations:\n",
    "`replaceWith0` - replaces all NAs with 0\n",
    "`geneInDf` - subseting data.frames by genename\n",
    "`geneDf` - extracting data.frames of each gene and respective annotations.\n",
    "`cadd.df` - extracts median CADD score of genes where multiple variants are present\n",
    "`rankByCaddScore` - for prioritizing genes by deleteriousness (scaled CADD score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following function reads in each output VCF file using the VariantAnnotation package available from Bioconductor. Install with `biocLite(\"VariantAnnotation\")`. `formatHgvs` (from the *myvariant* Bioconductor package) is a function that reads the genomic location and variant information from the VCF to create HGVS IDs which serve as a primary key for each variant. The function `getVariants` makes the queries to MyVariant.info to retrieve annotations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "found header lines for 3 ‘fixed’ fields: ALT, QUAL, FILTER \n",
      "found header lines for 24 ‘info’ fields: AC, AF, AN, BaseQRankSum, ClippingRankSum, DB, DP, DS, END, FS, HaplotypeScore, InbreedingCoeff, MLEAC, MLEAF, MQ, MQ0, MQRankSum, NEGATIVE_TRAIN_SITE, POSITIVE_TRAIN_SITE, QD, ReadPosRankSum, SOR, VQSLOD, culprit \n",
      "found header lines for 5 ‘geno’ fields: GT, AD, DP, GQ, PL \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Querying chunk 1 of 40\n",
      "Querying chunk 2 of 40\n"
     ]
    }
   ],
   "source": [
    "getVars <- function(vcf.file){\n",
    "    \n",
    "  vcf <- readVcf(vcf.file, genome=\"hg19\")\n",
    "  vcf <- vcf[isSNV(vcf)]\n",
    "  vars <- rowRanges(vcf)\n",
    "  vars <- as(vars, \"DataFrame\")\n",
    "  vars$query <- formatHgvs(vcf, \"snp\")\n",
    "  annotations <- getVariants(vars$query)\n",
    "  annotations[c('DP', 'FS', 'QD')] <- info(vcf)[c('DP', 'FS', 'QD')]\n",
    "    ## replace 'NA' with 0\n",
    "  annotations <- replaceWith0(annotations)\n",
    "  annotations\n",
    "}\n",
    "\n",
    "vars <- lapply(vcf.files, getVars)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'X_id'</li>\n",
       "\t<li>'query'</li>\n",
       "\t<li>'notfound'</li>\n",
       "\t<li>'dbsnp.allele_origin'</li>\n",
       "\t<li>'dbsnp.alleles'</li>\n",
       "\t<li>'dbsnp.alt'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'X_id'\n",
       "\\item 'query'\n",
       "\\item 'notfound'\n",
       "\\item 'dbsnp.allele_origin'\n",
       "\\item 'dbsnp.alleles'\n",
       "\\item 'dbsnp.alt'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'X_id'\n",
       "2. 'query'\n",
       "3. 'notfound'\n",
       "4. 'dbsnp.allele_origin'\n",
       "5. 'dbsnp.alleles'\n",
       "6. 'dbsnp.alt'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"X_id\"                \"query\"               \"notfound\"           \n",
       "[4] \"dbsnp.allele_origin\" \"dbsnp.alleles\"       \"dbsnp.alt\"          "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## check annotaion fields\n",
    "head(names(vars[[1]]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We then apply the function to each of the VCF files and do our filtering based on coverage and annotations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "filterDf <- function(df){\n",
    "  df <- subset(df, DP > 15 & FS < 30 & QD > 2)\n",
    "  df <- subset(df, cadd.consequence %in% c(\"NON_SYNONYMOUS\", \"STOP_GAINED\", \"STOP_LOST\", \n",
    "                                           \"CANONICAL_SPLICE\", \"SPLICE_SITE\"))\n",
    "  df <- subset(df, exac.af < 0.01)\n",
    "  df <- subset(df, sapply(dbnsfp.1000gp1.af, function(i) i < 0.01 ))\n",
    "  df\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "filtered.annotations <- lapply(vars, filterDf)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can then keep only genes that are mutated in each of the four patients."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>Var1</th><th scope=col>Freq</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>0</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>281</th><td>CDC27</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>293</th><td>CDON</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>388</th><td>CTBP2</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>391</th><td>CTDSP2</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>430</th><td>DHODH</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>538</th><td>FAM104B</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>599</th><td>FRG2C</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>693</th><td>GXYLT1</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>748</th><td>HYDIN</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>863</th><td>KIR3DL3</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>1021</th><td>MTCH2</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>1132</th><td>NULL</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>1300</th><td>POTED</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>1485</th><td>SETD8</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>1833</th><td>XKR3</td><td>4</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ll}\n",
       "  & Var1 & Freq\\\\\n",
       "\\hline\n",
       "\t1 & 0 & 4\\\\\n",
       "\t281 & CDC27 & 4\\\\\n",
       "\t293 & CDON & 4\\\\\n",
       "\t388 & CTBP2 & 4\\\\\n",
       "\t391 & CTDSP2 & 4\\\\\n",
       "\t430 & DHODH & 4\\\\\n",
       "\t538 & FAM104B & 4\\\\\n",
       "\t599 & FRG2C & 4\\\\\n",
       "\t693 & GXYLT1 & 4\\\\\n",
       "\t748 & HYDIN & 4\\\\\n",
       "\t863 & KIR3DL3 & 4\\\\\n",
       "\t1021 & MTCH2 & 4\\\\\n",
       "\t1132 & NULL & 4\\\\\n",
       "\t1300 & POTED & 4\\\\\n",
       "\t1485 & SETD8 & 4\\\\\n",
       "\t1833 & XKR3 & 4\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "        Var1 Freq\n",
       "1          0    4\n",
       "281    CDC27    4\n",
       "293     CDON    4\n",
       "388    CTBP2    4\n",
       "391   CTDSP2    4\n",
       "430    DHODH    4\n",
       "538  FAM104B    4\n",
       "599    FRG2C    4\n",
       "693   GXYLT1    4\n",
       "748    HYDIN    4\n",
       "863  KIR3DL3    4\n",
       "1021   MTCH2    4\n",
       "1132    NULL    4\n",
       "1300   POTED    4\n",
       "1485   SETD8    4\n",
       "1833    XKR3    4"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene.counts <- data.frame(table(unlist(lapply(filtered.annotations, function(i) unique(i$dbnsfp.genename)))))\n",
    "top.genes <- subset(gene.counts, Freq == 4)\n",
    "top.genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can prioritize the 14 genes that contain variants in each of the patients according to CADD (deleteriousness) score. rankByCaddScore extracts the average CADD scores of the variants in each gene per patient and ranks in descending order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in do.call(rbind, lapply(gene.list, function(i) geneDf(df.list, : error in evaluating the argument 'args' in selecting a method for function 'do.call': Error in lapply(gene.list, function(i) geneDf(df.list, i)) : \n  error in evaluating the argument 'X' in selecting a method for function 'lapply': Error: object 'top.genes' not found\n\n",
     "output_type": "error",
     "traceback": [
      "Error in do.call(rbind, lapply(gene.list, function(i) geneDf(df.list, : error in evaluating the argument 'args' in selecting a method for function 'do.call': Error in lapply(gene.list, function(i) geneDf(df.list, i)) : \n  error in evaluating the argument 'X' in selecting a method for function 'lapply': Error: object 'top.genes' not found\n\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in eval(expr, envir, enclos): object 'ranked' not found\n",
     "output_type": "error",
     "traceback": [
      "Error in eval(expr, envir, enclos): object 'ranked' not found\n"
     ]
    }
   ],
   "source": [
    "ranked <- rankByCaddScore(top.genes$Var1, filtered.annotations)\n",
    "ranked"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ng et al concluded DHODH is responsible for Miller Syndrome as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in unlist(x): error in evaluating the argument 'x' in selecting a method for function 'unlist': Error: object 'ranked' not found\n\n",
     "output_type": "error",
     "traceback": [
      "Error in unlist(x): error in evaluating the argument 'x' in selecting a method for function 'unlist': Error: object 'ranked' not found\n\n"
     ]
    }
   ],
   "source": [
    "queryMany(ranked$gene, scopes=\"symbol\", species=\"human\", fields=c(\"name\", \"pathway\", \"kegg\", \"MIM\", \"uniprot\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.2.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
